File: swift/532_K-diffPairsInAnArray.swift
class Solution {
    func findPairs(_ nums: [Int], _ k: Int) -> Int {
        var rollingNumCounts: [Int : Int] = [Int : Int]()
        var count: Int = 0
        for n in nums {
            var nCount = rollingNumCounts[n] ?? 0
            if k != 0 {
                if nCount == 0 { // first time seeing n so counting unique pairs
                    if rollingNumCounts[n - k] ?? 0 > 0 { count += 1 }
                    if rollingNumCounts[n + k] ?? 0 > 0 { count += 1 }
                }
            } else if nCount == 1 {
                count += 1
            }
            rollingNumCounts[n] = nCount > 0 ? nCount + 1 : 1
        }
        return count
    }
}

File: java/532_K-diffPairsInAnArray.java
class Solution {
    public int findPairs(int[] nums, int k) {
        HashMap<Integer, Integer> rollingNumCounts = new HashMap<Integer, Integer>();
        int count = 0;
        for (int n : nums) {
            int nCount = rollingNumCounts.getOrDefault(n, 0);
            if (k != 0) {
                if (nCount == 0) {
                    if (rollingNumCounts.getOrDefault(n - k, 0) > 0) ++count;
                    if (rollingNumCounts.getOrDefault(n + k, 0) > 0) ++count;
                }
            } else if (rollingNumCounts.getOrDefault(n, 0) == 1) 
                ++count;
            rollingNumCounts.put(n, nCount + 1);
        }
        return count;
    }
}
